src/states/common/base.rs

```diff
@@ -304,13 +318,52 @@ pub trait Base: Display {
         &mut self,
         addr: SocketAddr,
         msg: Bytes,
-        token: Token,
-        _outbox: &mut dyn EventBox,
+        msg_token: Token,
+        outbox: &mut dyn EventBox,
     ) -> Transition {
-        let log_ident = LogIdent::new(self);
-        self.network_service_mut()
-            .send_message_to_next_target(msg, token, addr, log_ident);
-        Transition::Stay
+        let next_target = self.network_service_mut().target_failed(msg_token, addr);
+        match next_target.failed_attempts {
+            0 => {
+                trace!(
+                    "{} - Sending message ID {} to {} failed - resending to {} now",
+                    self,
+                    msg_token,
+                    addr,
+                    next_target.addr
+                );
+                self.network_service_mut()
+                    .send_now(next_target.addr, msg, msg_token);
+                Transition::Stay
+            }
+            n if n < RESEND_MAX_ATTEMPTS => {
+                trace!(
+                    "{} - Sending message ID {} to {} failed - resending to {} in {:?}",
+                    self,
+                    msg_token,
+                    addr,
+                    next_target.addr,
+                    RESEND_DELAY
+                );
+                let timer_token = self.timer().schedule(RESEND_DELAY);
+                self.network_service_mut().send_later(
```

Comment:
would like to clarify that the total detection time of a failing peer via a particular message will then be: 
    RESEND_MAX_ATTEMPTS * (RESEND_DELAY + quic-p2p_detect_sending_failure_time) * number_of_failed_peers_at_the_same_time

Reply:
In case there is only one target, or if the number of targets is the same as the delivery group, then the total time to detect them offline is `RESEND_MAX_ATTEMPTS * (RESEND_DELAY + quic-p2p send failure time)`. If there are multiple targets and the delivery group is not all of them, the it gets more complicated. For example, let's say we send to 3 of 7. Two of them succeed and one fails. Then we pick next target to resend to, one which we haven't sent to before. If this succeeds then the message is successfully sent and we don't need to retry the previously failed one. So we would not detect them being offline at all. But that's OK, because there are always some other direct messages / swarm messages on which we would detect them.

Reply:
inside the function of `take_next_target` at https://github.com/maidsafe/routing/pull/2050/files#diff-79a9599b4b67afb5d69b448a0c38fedbR97 , 
the filter seems only pick the `failed` target as candidate for the `next target`.
that's why it is `number_of_failed_peers_at_the_same_time` mentioned in my formula.